// @daltonr/pathwrite-core/dist/index.js
export class PathEngine {
    activePath = null;
    pathStack = [];
    listeners = new Set();
    _isNavigating = false;
    subscribe(listener) {
        this.listeners.add(listener);
        return () => this.listeners.delete(listener);
    }
    // ---------------------------------------------------------------------------
    // Public API
    // ---------------------------------------------------------------------------
    start(path, initialData = {}) {
        this.assertPathHasSteps(path);
        return this._startAsync(path, initialData);
    }
    /** Starts a sub-path on top of the currently active path. Throws if no path is running. */
    startSubPath(path, initialData = {}) {
        this.requireActivePath();
        return this.start(path, initialData);
    }
    next() {
        const active = this.requireActivePath();
        return this._nextAsync(active);
    }
    previous() {
        const active = this.requireActivePath();
        return this._previousAsync(active);
    }
    /** Cancel is synchronous (no hooks). Returns a resolved Promise for API consistency. */
    cancel() {
        const active = this.requireActivePath();
        if (this._isNavigating)
            return Promise.resolve();
        const cancelledPathId = active.definition.id;
        const cancelledData = { ...active.data };
        if (this.pathStack.length > 0) {
            this.activePath = this.pathStack.pop() ?? null;
            this.emitStateChanged();
            return Promise.resolve();
        }
        this.activePath = null;
        this.emit({ type: "cancelled", pathId: cancelledPathId, data: cancelledData });
        return Promise.resolve();
    }
    setData(key, value) {
        const active = this.requireActivePath();
        active.data[key] = value;
        this.emitStateChanged();
        return Promise.resolve();
    }
    /** Jumps directly to the step with the given ID. Does not check guards or shouldSkip. */
    goToStep(stepId) {
        const active = this.requireActivePath();
        const targetIndex = active.definition.steps.findIndex((s) => s.id === stepId);
        if (targetIndex === -1) {
            throw new Error(`Step "${stepId}" not found in path "${active.definition.id}".`);
        }
        return this._goToStepAsync(active, targetIndex);
    }
    snapshot() {
        if (this.activePath === null) {
            return null;
        }
        const active = this.activePath;
        const step = this.getCurrentStep(active);
        const { steps } = active.definition;
        const stepCount = steps.length;
        return {
            pathId: active.definition.id,
            stepId: step.id,
            stepTitle: step.title,
            stepMeta: step.meta,
            stepIndex: active.currentStepIndex,
            stepCount,
            progress: stepCount <= 1 ? 1 : active.currentStepIndex / (stepCount - 1),
            steps: steps.map((s, i) => ({
                id: s.id,
                title: s.title,
                meta: s.meta,
                status: i < active.currentStepIndex ? "completed"
                    : i === active.currentStepIndex ? "current"
                        : "upcoming"
            })),
            isFirstStep: active.currentStepIndex === 0,
            isLastStep: active.currentStepIndex === stepCount - 1 &&
                this.pathStack.length === 0,
            nestingLevel: this.pathStack.length,
            isNavigating: this._isNavigating,
            canMoveNext: this.evaluateGuardSync(step.canMoveNext, active),
            canMovePrevious: this.evaluateGuardSync(step.canMovePrevious, active),
            data: { ...active.data }
        };
    }
    // ---------------------------------------------------------------------------
    // Private async helpers
    // ---------------------------------------------------------------------------
    async _startAsync(path, initialData) {
        if (this._isNavigating)
            return;
        if (this.activePath !== null) {
            this.pathStack.push(this.activePath);
        }
        this.activePath = {
            definition: path,
            currentStepIndex: 0,
            data: { ...initialData }
        };
        this._isNavigating = true;
        await this.skipSteps(1);
        if (this.activePath.currentStepIndex >= path.steps.length) {
            this._isNavigating = false;
            await this.finishActivePath();
            return;
        }
        this.emitStateChanged();
        try {
            this.applyPatch(await this.enterCurrentStep());
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _nextAsync(active) {
        if (this._isNavigating)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const step = this.getCurrentStep(active);
            if (await this.canMoveNext(active, step)) {
                this.applyPatch(await this.leaveCurrentStep(active, step));
                active.currentStepIndex += 1;
                await this.skipSteps(1);
                if (active.currentStepIndex >= active.definition.steps.length) {
                    this._isNavigating = false;
                    await this.finishActivePath();
                    return;
                }
                this.applyPatch(await this.enterCurrentStep());
            }
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _previousAsync(active) {
        if (this._isNavigating)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const step = this.getCurrentStep(active);
            if (await this.canMovePrevious(active, step)) {
                this.applyPatch(await this.leaveCurrentStep(active, step));
                active.currentStepIndex -= 1;
                await this.skipSteps(-1);
                if (active.currentStepIndex < 0) {
                    this._isNavigating = false;
                    await this.cancel();
                    return;
                }
                this.applyPatch(await this.enterCurrentStep());
            }
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _goToStepAsync(active, targetIndex) {
        if (this._isNavigating)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const currentStep = this.getCurrentStep(active);
            this.applyPatch(await this.leaveCurrentStep(active, currentStep));
            active.currentStepIndex = targetIndex;
            this.applyPatch(await this.enterCurrentStep());
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async finishActivePath() {
        const finished = this.requireActivePath();
        const finishedPathId = finished.definition.id;
        const finishedData = { ...finished.data };
        if (this.pathStack.length > 0) {
            this.activePath = this.pathStack.pop();
            const parent = this.activePath;
            const parentStep = this.getCurrentStep(parent);
            if (parentStep.onSubPathComplete) {
                const ctx = {
                    pathId: parent.definition.id,
                    stepId: parentStep.id,
                    data: { ...parent.data }
                };
                this.applyPatch(await parentStep.onSubPathComplete(finishedPathId, finishedData, ctx));
            }
            this.emit({
                type: "resumed",
                resumedPathId: parent.definition.id,
                fromSubPathId: finishedPathId,
                snapshot: this.snapshot()
            });
        }
        else {
            this.activePath = null;
            this.emit({ type: "completed", pathId: finishedPathId, data: finishedData });
        }
    }
    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------
    requireActivePath() {
        if (this.activePath === null) {
            throw new Error("No active path.");
        }
        return this.activePath;
    }
    assertPathHasSteps(path) {
        if (!path.steps || path.steps.length === 0) {
            throw new Error(`Path "${path.id}" must have at least one step.`);
        }
    }
    emit(event) {
        for (const listener of this.listeners) {
            listener(event);
        }
    }
    emitStateChanged() {
        this.emit({ type: "stateChanged", snapshot: this.snapshot() });
    }
    getCurrentStep(active) {
        return active.definition.steps[active.currentStepIndex];
    }
    applyPatch(patch) {
        if (patch && typeof patch === "object") {
            const active = this.activePath;
            if (active) {
                Object.assign(active.data, patch);
            }
        }
    }
    async skipSteps(direction) {
        const active = this.activePath;
        if (!active)
            return;
        while (active.currentStepIndex >= 0 &&
            active.currentStepIndex < active.definition.steps.length) {
            const step = active.definition.steps[active.currentStepIndex];
            if (!step.shouldSkip)
                break;
            const ctx = {
                pathId: active.definition.id,
                stepId: step.id,
                data: { ...active.data }
            };
            const skip = await step.shouldSkip(ctx);
            if (!skip)
                break;
            active.currentStepIndex += direction;
        }
    }
    async enterCurrentStep() {
        const active = this.activePath;
        if (!active)
            return;
        const step = this.getCurrentStep(active);
        if (!step.onEnter)
            return;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data }
        };
        return step.onEnter(ctx);
    }
    async leaveCurrentStep(active, step) {
        if (!step.onLeave)
            return;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data }
        };
        return step.onLeave(ctx);
    }
    async canMoveNext(active, step) {
        if (!step.canMoveNext)
            return true;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data }
        };
        return step.canMoveNext(ctx);
    }
    async canMovePrevious(active, step) {
        if (!step.canMovePrevious)
            return true;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data }
        };
        return step.canMovePrevious(ctx);
    }
    /**
     * Evaluates a guard function synchronously for inclusion in the snapshot.
     * If the guard is absent, returns `true`.
     * If the guard returns a `Promise`, returns `true` (optimistic default).
     */
    evaluateGuardSync(guard, active) {
        if (!guard)
            return true;
        const ctx = {
            pathId: active.definition.id,
            stepId: this.getCurrentStep(active).id,
            data: { ...active.data }
        };
        const result = guard(ctx);
        if (typeof result === "boolean")
            return result;
        // Async guard — default to true (optimistic); the engine will enforce the real result on navigation.
        return true;
    }
}
