// @daltonr/pathwrite-core/dist/index.js
export class PathEngine {
    activePath = null;
    pathStack = [];
    listeners = new Set();
    _isNavigating = false;
    subscribe(listener) {
        this.listeners.add(listener);
        return () => this.listeners.delete(listener);
    }
    // ---------------------------------------------------------------------------
    // Public API
    // ---------------------------------------------------------------------------
    start(path, initialData = {}) {
        this.assertPathHasSteps(path);
        return this._startAsync(path, initialData);
    }
    /**
     * Starts a sub-path on top of the currently active path. Throws if no path
     * is running.
     *
     * @param path        The sub-path definition to start.
     * @param initialData Data to seed the sub-path with.
     * @param meta        Optional correlation object returned unchanged to the
     *                    parent step's `onSubPathComplete` / `onSubPathCancel`
     *                    hooks. Use to identify which collection item triggered
     *                    the sub-path without embedding that information in the
     *                    sub-path's own data.
     */
    startSubPath(path, initialData = {}, meta) {
        this.requireActivePath();
        return this._startAsync(path, initialData, meta);
    }
    next() {
        const active = this.requireActivePath();
        return this._nextAsync(active);
    }
    previous() {
        const active = this.requireActivePath();
        return this._previousAsync(active);
    }
    /** Cancel is synchronous for top-level paths (no hooks). Sub-path cancellation
     *  is async when an `onSubPathCancel` hook is present. Returns a Promise for
     *  API consistency. */
    cancel() {
        const active = this.requireActivePath();
        if (this._isNavigating)
            return Promise.resolve();
        const cancelledPathId = active.definition.id;
        const cancelledData = { ...active.data };
        const cancelledMeta = active.subPathMeta;
        if (this.pathStack.length > 0) {
            return this._cancelSubPathAsync(cancelledPathId, cancelledData, cancelledMeta);
        }
        this.activePath = null;
        this.emit({ type: "cancelled", pathId: cancelledPathId, data: cancelledData });
        return Promise.resolve();
    }
    setData(key, value) {
        const active = this.requireActivePath();
        active.data[key] = value;
        this.emitStateChanged();
        return Promise.resolve();
    }
    /** Jumps directly to the step with the given ID. Does not check guards or shouldSkip. */
    goToStep(stepId) {
        const active = this.requireActivePath();
        const targetIndex = active.definition.steps.findIndex((s) => s.id === stepId);
        if (targetIndex === -1) {
            throw new Error(`Step "${stepId}" not found in path "${active.definition.id}".`);
        }
        return this._goToStepAsync(active, targetIndex);
    }
    /**
     * Jumps directly to the step with the given ID, but first checks the
     * direction-appropriate guard on the current step:
     * - Going forward  → checks `canMoveNext`
     * - Going backward → checks `canMovePrevious`
     *
     * If the guard blocks, navigation does not occur and `stateChanged` is still
     * emitted (so the UI can react). `shouldSkip` is not evaluated.
     * Throws if the target step ID does not exist.
     */
    goToStepChecked(stepId) {
        const active = this.requireActivePath();
        const targetIndex = active.definition.steps.findIndex((s) => s.id === stepId);
        if (targetIndex === -1) {
            return Promise.reject(new Error(`Step "${stepId}" not found in path "${active.definition.id}".`));
        }
        if (targetIndex === active.currentStepIndex)
            return Promise.resolve();
        return this._goToStepCheckedAsync(active, targetIndex);
    }
    snapshot() {
        if (this.activePath === null) {
            return null;
        }
        const active = this.activePath;
        const step = this.getCurrentStep(active);
        const { steps } = active.definition;
        const stepCount = steps.length;
        return {
            pathId: active.definition.id,
            stepId: step.id,
            stepTitle: step.title,
            stepMeta: step.meta,
            stepIndex: active.currentStepIndex,
            stepCount,
            progress: stepCount <= 1 ? 1 : active.currentStepIndex / (stepCount - 1),
            steps: steps.map((s, i) => ({
                id: s.id,
                title: s.title,
                meta: s.meta,
                status: i < active.currentStepIndex ? "completed"
                    : i === active.currentStepIndex ? "current"
                        : "upcoming"
            })),
            isFirstStep: active.currentStepIndex === 0,
            isLastStep: active.currentStepIndex === stepCount - 1 &&
                this.pathStack.length === 0,
            nestingLevel: this.pathStack.length,
            isNavigating: this._isNavigating,
            canMoveNext: this.evaluateGuardSync(step.canMoveNext, active),
            canMovePrevious: this.evaluateGuardSync(step.canMovePrevious, active),
            validationMessages: this.evaluateValidationMessagesSync(step.validationMessages, active),
            data: { ...active.data }
        };
    }
    // ---------------------------------------------------------------------------
    // Private async helpers
    // ---------------------------------------------------------------------------
    async _startAsync(path, initialData, subPathMeta) {
        if (this._isNavigating)
            return;
        if (this.activePath !== null) {
            this.pathStack.push(this.activePath);
        }
        this.activePath = {
            definition: path,
            currentStepIndex: 0,
            data: { ...initialData },
            visitedStepIds: new Set(),
            subPathMeta
        };
        this._isNavigating = true;
        await this.skipSteps(1);
        if (this.activePath.currentStepIndex >= path.steps.length) {
            this._isNavigating = false;
            await this.finishActivePath();
            return;
        }
        this.emitStateChanged();
        try {
            this.applyPatch(await this.enterCurrentStep());
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _nextAsync(active) {
        if (this._isNavigating)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const step = this.getCurrentStep(active);
            if (await this.canMoveNext(active, step)) {
                this.applyPatch(await this.leaveCurrentStep(active, step));
                active.currentStepIndex += 1;
                await this.skipSteps(1);
                if (active.currentStepIndex >= active.definition.steps.length) {
                    this._isNavigating = false;
                    await this.finishActivePath();
                    return;
                }
                this.applyPatch(await this.enterCurrentStep());
            }
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _previousAsync(active) {
        if (this._isNavigating)
            return;
        // No-op when already on the first step of a top-level path.
        // Sub-paths still cancel/pop back to the parent when previous() is called
        // on their first step (the currentStepIndex < 0 branch below handles that).
        if (active.currentStepIndex === 0 && this.pathStack.length === 0)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const step = this.getCurrentStep(active);
            if (await this.canMovePrevious(active, step)) {
                this.applyPatch(await this.leaveCurrentStep(active, step));
                active.currentStepIndex -= 1;
                await this.skipSteps(-1);
                if (active.currentStepIndex < 0) {
                    this._isNavigating = false;
                    await this.cancel();
                    return;
                }
                this.applyPatch(await this.enterCurrentStep());
            }
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _goToStepAsync(active, targetIndex) {
        if (this._isNavigating)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const currentStep = this.getCurrentStep(active);
            this.applyPatch(await this.leaveCurrentStep(active, currentStep));
            active.currentStepIndex = targetIndex;
            this.applyPatch(await this.enterCurrentStep());
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _goToStepCheckedAsync(active, targetIndex) {
        if (this._isNavigating)
            return;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const currentStep = this.getCurrentStep(active);
            const goingForward = targetIndex > active.currentStepIndex;
            const allowed = goingForward
                ? await this.canMoveNext(active, currentStep)
                : await this.canMovePrevious(active, currentStep);
            if (allowed) {
                this.applyPatch(await this.leaveCurrentStep(active, currentStep));
                active.currentStepIndex = targetIndex;
                this.applyPatch(await this.enterCurrentStep());
            }
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async _cancelSubPathAsync(cancelledPathId, cancelledData, cancelledMeta) {
        // Pop the stack BEFORE emitting so snapshot() always reflects the parent
        // path (which has a valid currentStepIndex) rather than the cancelled
        // sub-path (which may have currentStepIndex = -1).
        this.activePath = this.pathStack.pop() ?? null;
        this._isNavigating = true;
        this.emitStateChanged();
        try {
            const parent = this.activePath;
            if (parent) {
                const parentStep = this.getCurrentStep(parent);
                if (parentStep.onSubPathCancel) {
                    const ctx = {
                        pathId: parent.definition.id,
                        stepId: parentStep.id,
                        data: { ...parent.data },
                        isFirstEntry: !parent.visitedStepIds.has(parentStep.id)
                    };
                    this.applyPatch(await parentStep.onSubPathCancel(cancelledPathId, cancelledData, ctx, cancelledMeta));
                }
            }
            this._isNavigating = false;
            this.emitStateChanged();
        }
        catch (err) {
            this._isNavigating = false;
            this.emitStateChanged();
            throw err;
        }
    }
    async finishActivePath() {
        const finished = this.requireActivePath();
        const finishedPathId = finished.definition.id;
        const finishedData = { ...finished.data };
        const finishedMeta = finished.subPathMeta;
        if (this.pathStack.length > 0) {
            this.activePath = this.pathStack.pop();
            const parent = this.activePath;
            const parentStep = this.getCurrentStep(parent);
            if (parentStep.onSubPathComplete) {
                const ctx = {
                    pathId: parent.definition.id,
                    stepId: parentStep.id,
                    data: { ...parent.data },
                    isFirstEntry: !parent.visitedStepIds.has(parentStep.id)
                };
                this.applyPatch(await parentStep.onSubPathComplete(finishedPathId, finishedData, ctx, finishedMeta));
            }
            this.emit({
                type: "resumed",
                resumedPathId: parent.definition.id,
                fromSubPathId: finishedPathId,
                snapshot: this.snapshot()
            });
        }
        else {
            this.activePath = null;
            this.emit({ type: "completed", pathId: finishedPathId, data: finishedData });
        }
    }
    // ---------------------------------------------------------------------------
    // Private helpers
    // ---------------------------------------------------------------------------
    requireActivePath() {
        if (this.activePath === null) {
            throw new Error("No active path.");
        }
        return this.activePath;
    }
    assertPathHasSteps(path) {
        if (!path.steps || path.steps.length === 0) {
            throw new Error(`Path "${path.id}" must have at least one step.`);
        }
    }
    emit(event) {
        for (const listener of this.listeners) {
            listener(event);
        }
    }
    emitStateChanged() {
        this.emit({ type: "stateChanged", snapshot: this.snapshot() });
    }
    getCurrentStep(active) {
        return active.definition.steps[active.currentStepIndex];
    }
    applyPatch(patch) {
        if (patch && typeof patch === "object") {
            const active = this.activePath;
            if (active) {
                Object.assign(active.data, patch);
            }
        }
    }
    async skipSteps(direction) {
        const active = this.activePath;
        if (!active)
            return;
        while (active.currentStepIndex >= 0 &&
            active.currentStepIndex < active.definition.steps.length) {
            const step = active.definition.steps[active.currentStepIndex];
            if (!step.shouldSkip)
                break;
            const ctx = {
                pathId: active.definition.id,
                stepId: step.id,
                data: { ...active.data },
                isFirstEntry: !active.visitedStepIds.has(step.id)
            };
            const skip = await step.shouldSkip(ctx);
            if (!skip)
                break;
            active.currentStepIndex += direction;
        }
    }
    async enterCurrentStep() {
        const active = this.activePath;
        if (!active)
            return;
        const step = this.getCurrentStep(active);
        const isFirstEntry = !active.visitedStepIds.has(step.id);
        // Mark as visited before calling onEnter so re-entrant calls see the
        // correct isFirstEntry value.
        active.visitedStepIds.add(step.id);
        if (!step.onEnter)
            return;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data },
            isFirstEntry
        };
        return step.onEnter(ctx);
    }
    async leaveCurrentStep(active, step) {
        if (!step.onLeave)
            return;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data },
            isFirstEntry: !active.visitedStepIds.has(step.id)
        };
        return step.onLeave(ctx);
    }
    async canMoveNext(active, step) {
        if (!step.canMoveNext)
            return true;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data },
            isFirstEntry: !active.visitedStepIds.has(step.id)
        };
        return step.canMoveNext(ctx);
    }
    async canMovePrevious(active, step) {
        if (!step.canMovePrevious)
            return true;
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data },
            isFirstEntry: !active.visitedStepIds.has(step.id)
        };
        return step.canMovePrevious(ctx);
    }
    /**
     * Evaluates a guard function synchronously for inclusion in the snapshot.
     * If the guard is absent, returns `true`.
     * If the guard returns a `Promise`, returns `true` (optimistic default).
     */
    evaluateGuardSync(guard, active) {
        if (!guard)
            return true;
        const step = this.getCurrentStep(active);
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data },
            isFirstEntry: !active.visitedStepIds.has(step.id)
        };
        const result = guard(ctx);
        if (typeof result === "boolean")
            return result;
        return true;
    }
    /**
     * Evaluates a validationMessages function synchronously for inclusion in the snapshot.
     * If the hook is absent, returns `[]`.
     * If the hook returns a `Promise`, returns `[]` (async hooks are not supported in snapshots).
     */
    evaluateValidationMessagesSync(fn, active) {
        if (!fn)
            return [];
        const step = this.getCurrentStep(active);
        const ctx = {
            pathId: active.definition.id,
            stepId: step.id,
            data: { ...active.data },
            isFirstEntry: !active.visitedStepIds.has(step.id)
        };
        const result = fn(ctx);
        if (Array.isArray(result))
            return result;
        return [];
    }
}
